scripts/codegen/modules/devicedeclare.py

```diff
@@ -0,0 +1,391 @@
+        codegen.outl(device_info)
+    #
+    # device init
+    codegen.outl(_device_template_substitute(_device_and_api_init_tmpl,
+                                             device_id, preset))
+    return True
+
+##
+# @param device_configs
+#   A list of configuration variables for device instantiation.
+#   (e.g. ['CONFIG_SPI_0', 'CONFIG_SPI_1'])
+# @param driver_names
+#   A list of driver names for device instantiation. The list shall be ordered
+#   as the list of device configs. Watch for the '"' of the driver names.
+#   (e.g. ['"SPI_0"', '"SPI_1"'])
+# @param device_inits
+#   A list of device initialisation functions or a one single function. The
+#   list shall be ordered as the list of device configs.
+#   (e.g. 'spi_stm32_init')
+# @param device_levels
+#   A list of driver initialisation levels or one single level definition. The
+#   list shall be ordered as the list of device configs.
+#   (e.g. 'PRE_KERNEL_1')
+# @param device_prios
+#   A list of driver initialisation priorities or one single priority
+#   definition. The list shall be ordered as the list of device configs.
+#   (e.g. 32)
+# @param device_api
+#   Identifier of the device api.
+#   (e.g. 'spi_stm32_driver_api')
+# @param device_info
+#   Device info template for device driver config, data and interrupt
+#   initialisation.
+#
+def device_declare_multi(device_configs,
+                         driver_names,
+                         device_inits,
+                         device_levels,
+                         device_prios,
+                         device_api,
+                         device_info):
+    devices_declared = []
+    for i, device_config in enumerate(device_configs):
+        driver_name = driver_names[i]
+        if isinstance(device_inits, str):
+            device_init = device_inits
+        else:
+            try:
+                device_init = device_inits[i]
+            except:
+                device_init = device_inits
+        if isinstance(device_levels, str):
+            device_level = device_levels
+        else:
+            try:
+                device_level = device_levels[i]
+            except:
+                device_level = device_levels
+        if isinstance(device_prios, str):
+            device_prio = device_prios
+        else:
+            try:
+                device_prio = device_prios[i]
+            except:
+                device_prio = device_prios
+
+        device_declared = device_declare_single(device_config,
+                                                driver_name,
+                                                device_init,
+                                                device_level,
+                                                device_prio,
+                                                device_api,
+                                                device_info)
+        devices_declared.append(device_declared)
+
+    if True not in devices_declared:
```

Comment:
> This is not an error.

You may differentiate between the functions:

- devicedeclare.device_declare_multi - For this function it is an **ERROR**. If you don´t instantiate a device then the driver should not have been foreseen for code generation/compilation.
- devicedeclare.device_declare_single: Here it is **NO ERROR**. Therefor this function does not throw an exception.
- devicedeclare.device_declare: Here it is **NO ERROR**. The usage of device_declare_multi in the implementation seems to be wrong. device_declare should be based on device_declare_single - which was recognized also in another comment.

Reply:
What is the purpose of device_declare_multi then?
Once again, using 'compatible' to parse the database you should not face this kind of issue: generation should not proceed if no instance with the compatible is available in the database.

Reply:
> generation should not proceed if no instance with the compatible is available in the database.

As you might see generation does not proceed. An exception is raised.

Reply:
Aim of the comment was to illustrate that you'd face this kind of error using 'compatible'.

Reply:
I assume you think I would not face this type of error if I would use 'compatible' instead of 'label'. Re-iterating the same again - please see the comments about 'label' vs. 'comaptible'.